python/lsst/image_cutout_backend/_image_cutout_backend.py

```diff
@@ -0,0 +1,412 @@
+    ):
+        self.butler = butler
+        self.projection_finder = projection_finder
+        self.output_root = ResourcePath(output_root, forceAbsolute=True, forceDirectory=True)
+        self.filename_template = filename_template
+        self.temporary_root = (
+            ResourcePath(temporary_root, forceDirectory=False) if temporary_root is not None else None
+        )
+
+    butler: Butler
+    """Butler client used to read cutouts (`Butler`).
+    """
+
+    projection_finder: ProjectionFinder
+    """Object that obtains the WCS and bounding box for butler datasets of
+    different types (`ProjectionFinder`).
+    """
+
+    output_root: ResourcePath
+    """Root path that extracted cutouts are written to (`ResourcePath`).
+    """
+
+    filename_template: str
+    """Template for output file names, not including the ``output_root``
+    prefix or the file type extension.
+    """
+
+    temporary_root: Optional[ResourcePath]
+    """Local filesystem root to write files to before they are transferred to
+    ``output_root``
+    """
+
+    def process_ref(
+        self, stencil: SkyStencil, ref: DatasetRef, *, mask_plane: Optional[str] = "STENCIL"
+    ) -> ResourcePath:
+        """Extract and write a cutout from a fully-resolved `DatasetRef`.
+
+        Parameters
+        ----------
+        stencil : `SkyStencil`
+            Definition of the cutout region, in sky coordinates.
+        ref : `DatasetRef`
+            Fully-resolved reference to the dataset to obtain the cutout from.
+            Must have ``DatasetRef.id`` not `None` (use `read_search` instead
+            when this is not the case).  Need not have an expanded data ID.
+            May represent an image-like dataset component.
+        mask : `str`, optional
+            If not `None`, set this mask plane in the extracted cutout showing
+            the approximate stencil region.  Does nothing if the image type
+            does not have a mask plane.  Defaults to ``STENCIL``.
+
+        Returns
+        -------
+        uri : `ResourcePath`
+            Full path to the extracted cutout.
+        """
+        read_result = self.read_ref(stencil, ref)
+        if mask_plane is not None:
+            read_result.mask(mask_plane)
+        return self.write_fits(read_result)
+
+    def process_uuid(
+        self,
+        stencil: SkyStencil,
+        uuid: UUID,
+        *,
+        component: Optional[str] = None,
+        mask_plane: Optional[str] = "STENCIL",
+    ) -> ResourcePath:
+        """Extract and write a cutout from a dataset identified by its UUID.
+
+        Parameters
+        ----------
+        stencil : `SkyStencil`
+            Definition of the cutout region, in sky coordinates.
+        uuid : `UUID`
+            Unique ID of the dataset to read from.
+        component : `str`, optional
+            If not `None` (default), read this component instead of the
+            composite dataset.
```

Comment:
I don't understand this parameter. Doesn't the UUID uniquely specify the dataset to use as the source?

Reply:
The UUID uniquely specifies the dataset from the Registry point of view. The datastore can split a single python type into multiple files (for example storing the image plane and variance plane as separate files). To uniquely access a specific file in datastore you might need the UUID and component.

Reply:
Ah, thanks. I think my question is then how do we anticipate this information being plumbed through DataLink and ObsTAP? The design here is that the cutout frontend doesn't know anything about the structure of identifiers; they're just opaque identifiers that it gets from an SIA or ObsTAP query and that it passes to the backend. So if we need something besides the UUID, we're back to needing a serialization format that SIA and ObsTAP will use so that the cutout service gets a serialized identifier that can be passed to the backend.

Reply:
ObsTAP and data link don't know about composite datasets so SODA isn't going to know about it either unless some how there's a way for a SODA user to ask for a cutout of the variance plane without wanting the image plane.

Reply:
Ah, okay, so basically I should ignore this parameter because the cutout service probably won't use it (at least unless we develop some local extension out of need later on)?

Reply:
Yeah, I think you can ignore this for now.  It could be a useful extension for users who want to get a lot of cutouts and are concerned about the total overall size, because the common case (just getting the main image plane) can save a factor of 2 and change, but not any kind of immediate concern.